presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java

```diff
@@ -0,0 +1,246 @@
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.Identifier;
+import com.facebook.presto.sql.tree.LambdaArgumentDeclaration;
+import com.facebook.presto.sql.tree.LambdaExpression;
+import com.facebook.presto.sql.tree.Literal;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.NodeLocation;
+import com.facebook.presto.sql.tree.NotExpression;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.SearchedCaseExpression;
+import com.facebook.presto.sql.tree.WhenClause;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+public class SimplifyArrayOperationsRewriter
+{
+    private SimplifyArrayOperationsRewriter() {}
+
+    public static Expression simplifyArrayOperations(Expression expression, Session session)
+    {
+        if (SystemSessionProperties.isSimplifyArrayOperations(session)) {
+            return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression);
+        }
+
+        return expression;
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private static boolean isConstant(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isConstant(((Cast) expression).getExpression());
+            }
+
+            return expression instanceof Literal;
+        }
+
+        private static boolean isZero(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isZero(((Cast) expression).getExpression());
+            }
+
+            return expression instanceof LongLiteral && ((LongLiteral) expression).getValue() == 0;
+        }
+
+        private static boolean isCardinalityOfFilter(Expression expression)
+        {
+            if (expression instanceof FunctionCall) {
+                FunctionCall functionCall = (FunctionCall) expression;
+                return functionCall.getName().toString().equals("cardinality") &&
+                        functionCall.getArguments().size() == 1 &&
+                        functionCall.getArguments().get(0) instanceof FunctionCall &&
+                        ((FunctionCall) (functionCall.getArguments().get(0))).getName().toString().equals("filter");
+            }
+
+            return false;
+        }
+
+        private static boolean isSimpleComaparison(ComparisonExpression comparisonExpression)
+        {
+            switch (comparisonExpression.getOperator()) {
+                case EQUAL:
+                case GREATER_THAN:
+                case LESS_THAN:
+                case NOT_EQUAL:
+                case GREATER_THAN_OR_EQUAL:
+                case LESS_THAN_OR_EQUAL:
+                    return true;
+            }
+
+            return false;
+        }
+
+        private Expression simplifyCardinalityOfFilterComparedToZero(
```

Comment:
I think it might be better if this method is written in the following layout:

```
- create case statement
- create inputFunction using the above expression
- create outputFunction
- create reduce function
```

Currently it was somewhat twisted so the logic is not as obvious.

Reply:
Yeah I was thinking about it but it is too much code bloat which I dislike very much. I'm not actually not a fan of this kind of rewriting. It should be string/unparse based.

Reply:
I think it's pretty obvious and I put some comments. So I will leave it like this for now.